src/mercury_engine_data_structures/formats/bcskla.py

```diff
@@ -0,0 +1,167 @@
+    )
+
+    def __init__(self):
+        super().__init__(self.SUBCON)
+
+    def _decode(self, obj, context, path):
+        res = ListContainer()
+        for i in range(obj.count):
+            res.append(
+                Container(
+                    time = obj.timings[i],
+                    value = obj["values"][i].value,
+                    derivative = obj["values"][i].derivative,
+                )
+            )
+
+        return res
+
+    def _encode(self, obj, context, path):
+        res = Container(
+            timing_type = 0 if obj[-1].time > 0xFF else 8,
+            count = len(obj),
+            timings = ListContainer([v.time for v in obj]),
+            values = ListContainer([Container(value=v.value, derivative=v.derivative) for v in obj])
+        )
+
+        return res
+
+
+
+KeyFramedValues_Dread = FocusedSeq(
+    "data",
+    data = DreadKFVAdapter(),
+    _new_eof=Tell,
+    _new_kfv=Computed(_update_next_kfv)
+)
+
+# if data is keyframed, it builds to { count = c, data = [ { timing = t, values = [a, b, c] }, ... ] }
+# if data is constant, it builds to { count = 0, data = { value = x } }
+KeyFramedValues_SR = Struct(
+    count=Rebuild(Int16ul, lambda ctx: 0 if isinstance(ctx.data, float) else len(ctx.data)),
+    _data_type = Const(2, Int16ul),
+    data = IfThenElse(
+        this.count == 0,
+        FocusedSeq(
+            "value",
+            timing = Const(0, Int32ul),
+            value = Float
+        ),
+        # guessing on left/right_derivative. they are almost always the same value.
+        Array(this.count, Struct(timing=Float, value=Float, left_derivative=Float, right_derivative=Float))
+    ),
+    _new_eof=Tell,
+    _new_kfv=Computed(_update_next_kfv)
+
+)
+
+BoneTrack_Dread = Struct(
+    # guess. almost always the bone names but some values still don't parse after dumping all model data >:(
+    bone_name=PropertyEnumDoubleUnsafe,
+    flags=Rebuild(Int32ul, _rebuild_flags),
+    _offset=Tell,
+    data=Array(9, IfThenElse(
+        # if the flag for an index is zero, the value is keyframed
+        2 ** this._index & this.flags == 0,
+        Float,
+        FocusedSeq(
+            "kfv",
+            # offsets are relative to immediately after the flag field
+            off = Rebuild(Int32ul, lambda ctx: ctx._._._next_kfv_offset - ctx._._offset),
+            kfv = Pointer(this.off + this._._offset, KeyFramedValues_Dread)
+        )
+    ))
+)
+
+BoneTrack_SR = Struct(
+    bone_hash=Int32ul, # TODO have a property CrcAdapter for sr lmao
+    data=Array(9, FocusedSeq(
+        "kfv",
+        off = Rebuild(Int32ul, lambda ctx: 0 if isinstance(ctx.kfv, float) else ctx._._._next_kfv_offset),
+        kfv = IfThenElse(
+            this.off > 0,
+            Pointer(this.off, KeyFramedValues_SR),
+            # if offset is 0, this is the default value pos/rot=0, scale=1
+            IfThenElse(lambda ctx: ctx._index < 6, Computed(0.0), Computed(1.0))
```

Comment:
a comment somewhere indicating that `data` is always pos_xyz, rot_xyz, scale_xyz in order would be really helpful for understanding the format

Reply:
actually, better yet, bake it into the format. instead of an array of 9 entries, do like uhh
```py
def KFV_SR(default: float):
    return FocusedSeq(
        "kfv",
        off = Rebuild(Int32ul, lambda ctx: 0 if isinstance(ctx.kfv, float) else ctx._._._next_kfv_offset),
        kfv = IfThenElse(
            this.off > 0,
            Pointer(this.off, KeyFramedValues_SR),
            Computed(default)
        )
    )

def Vec3KFV_SR(default: float):
    return Struct(
        x=KFV_SR(default),
        y=KFV_SR(default),
        z=KFV_SR(default),
    )

BoneTrack_SR = Struct(
    bone_hash=Int32ul,
    position=Vec3KFV_SR(0.0),
    rotation=Vec3KFV_SR(0.0),
    scale=Vec3KFV_SR(1.0),
)
```

for dread, maybe just use an adapter to transform the array into these. hell maybe it's easier to use an adapter for both idk. but I think it'll be waaay easier to work with the files if the data here is properly labeled